src/t_cose_sign_sign.c

```diff
@@ -0,0 +1,277 @@
+/*
+ * t_cose_sign_sign.c
+ *
+ * Copyright (c) 2018-2022, Laurence Lundblade. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * See BSD-3-Clause license in README.md
+ */
+
+#include "t_cose/t_cose_sign_sign.h"
+#include "qcbor/qcbor.h"
+#include "t_cose/t_cose_signature_sign.h"
+#include "t_cose/t_cose_parameters.h"
+
+
+/**
+ * \file t_cose_sign_sign.c
+ *
+ * \brief This implements t_cose signing
+ *
+ * Stack usage to sign is dependent on the signing alg and key size
+ * and type of hash implementation. t_cose_sign1_finish() is the main
+ * user of stack It is 384 for \ref COSE_ALGORITHM_ES256 and 778 for
+ * \ref COSE_ALGORITHM_ES512.
+ */
+
+
+/*
+ * Pubilc Function. See t_cose_sign_sign.h
+ */
+enum t_cose_err_t
+t_cose_sign_encode_start(struct t_cose_sign_sign_ctx *me,
+                         bool                         payload_is_detached,
+                         QCBOREncodeContext          *cbor_encode_ctx)
+{
+    enum t_cose_err_t                 return_value;
+    const struct t_cose_header_param *params_vector[3];
+    int                               vector_index;
+    struct t_cose_signature_sign     *signer;
+
+    signer = me->signers;
+    if(signer == NULL) {
+        /* No signers configured. */
+        return_value = 888;
+        goto Done;
+    }
+
+    vector_index = 0;
+    if(me->option_flags & T_COSE_OPT_COSE_SIGN1) {
+
+        /* For a COSE_Sign1, the header parameters go in the
+         * main body header parameter section, not in the
+         * signatures. Ask the first sigher for the header
+         * parameters it wants to output. */
+        (signer->h_callback)(signer, &params_vector[0]);
+        vector_index++;
+        if(signer->next_in_list != NULL) {
+            /* In COSE_Sign1 mode, but too many signers configured.*/
+            return_value = 999;
+            goto Done;
+        }
+    }
+    params_vector[vector_index] = me->added_body_parameters;
+    vector_index++;
+    params_vector[vector_index] = NULL;
+    /* --- parameters are now all in params_vector -- */
+
+
+    /* --- Add the CBOR tag indicating COSE_Sign1 --- */
+    if(!(me->option_flags & T_COSE_OPT_OMIT_CBOR_TAG)) {
```

Comment:
The problem with this if-clause is that it does not create COSE_SIGN-tagged structures. Change it to:

> 
>     if(!(me->option_flags & T_COSE_OPT_OMIT_CBOR_TAG)) {
>         if(me->option_flags & T_COSE_OPT_COSE_SIGN1) {
>             QCBOREncode_AddTag(cbor_encode_ctx, CBOR_TAG_COSE_SIGN1);
>         } else
>         {
>             QCBOREncode_AddTag(cbor_encode_ctx, CBOR_TAG_COSE_SIGN);
>         }
>     }